pandas/core/indexes/api.py

```diff
@@ -157,23 +158,43 @@ def _get_combined_index(
         index = ensure_index(index)
 
     if sort:
-        try:
-            array_sorted = safe_sort(index)
-            array_sorted = cast(np.ndarray, array_sorted)
-            if isinstance(index, MultiIndex):
-                index = MultiIndex.from_tuples(array_sorted, names=index.names)
-            else:
-                index = Index(array_sorted, name=index.name)
-        except TypeError:
-            pass
-
+        index = safe_sort_index(index)
     # GH 29879
     if copy:
         index = index.copy()
 
     return index
 
 
+def safe_sort_index(index: Index) -> Index:
+    """
+    Returns the sorted index
+
+    We keep the dtypes and the name attributes.
+
+    Parameters
+    ----------
+    index : an Index
+
+    Returns
+    -------
+    Index
+    """
+    try:
+        array_sorted = safe_sort(index)
+        array_sorted = cast(np.ndarray, array_sorted)
+        if isinstance(index, MultiIndex):
+            index = MultiIndex.from_tuples(
+                array_sorted, names=index.names, dtypes=index.dtypes
+            )
+        else:
+            index = Index(array_sorted, name=index.name, dtype=index.dtype)
+    except TypeError:
```

Comment:
maybe only have the safe_sort call in the try block?

Reply:
A bit tricky, since array_sorted does not exist in the except block and we don't want to caste the Index again to an Index?

Reply:
with an `else` block we would only cast if the safe_sort succeeded? (i.e. Index values are unique)

Reply:
Thanks very much. I am always forgetting that...

Changed the code